apps/dgst.c

```diff
@@ -290,14 +293,16 @@ int dgst_main(int argc, char **argv)
              */
             goto end;
         }
-        type = EVP_PKEY_get_id(sigkey);
-        if (type == EVP_PKEY_ED25519 || type == EVP_PKEY_ED448) {
-            /*
-             * We implement PureEdDSA for these which doesn't have a separate
-             * digest, and only supports one shot.
-             */
-            BIO_printf(bio_err, "Key type not supported for this operation\n");
-            goto end;
+        {
+            char def_md[80];
+
+            if (EVP_PKEY_get_default_digest_name(sigkey, def_md,
+                                                 sizeof(def_md)) == 2
+                    && strcmp(def_md, "UNDEF") == 0)
```

Comment:
I think we should have an API that returns boolean on whether or not this is a one-shot algorithm or this code will get copied into all the apps out there. I'd rather have a better option than a string compare.

Tackle that if you wish - I'll not block approving on that basis - but I would prefer it.

Reply:
I think this is out of scope for this PR.

EVP_PKEY knows about the keymanager which can query the 'madatory digest" param.
A better test would be to query if if supports signature updates, but this requires a ctx set up for signing in order to determine if the dispatch methods are available. This could get the PR stuck for quite a while, so it shoul dnot be done here..

Reply:
Maybe raise an issue to create an EVP_PKEY_supports_one_shot_digest function?
Wouldn't want this to get lost in the frenetic mess.

Reply:
Raised https://github.com/openssl/openssl/issues/26700